cupyx/scipy/special/_stats_distributions.py

```diff
@@ -32,6 +32,53 @@
     ''')
 
 
+log_ndtr_definition = """
+
+#define NPY_SQRT1_2   0.707106781186547524400844362104849039  /* 1/sqrt(2) */
+
+template <typename T>
+static __device__ T log_ndtr(T x)
+{
+    T t = x * NPY_SQRT1_2;
+    if (x < -1.0) {
+        return log(erfcx(-t) / 2) - t * t;
+    } else {
+        return log1p(-erfc(t) / 2);
+    }
+}
```

Comment:
We can change this algorithm to `log(ndtr)`. But I see scipy has used this algorithm, due to numerical precision issues. Ref: https://github.com/scipy/scipy/pull/15187. Is this okay? Thanks

Reply:
Using this formula is okay! BTW, would you also implement single precision version using `logf`, `log1pf`, `erfcf` and `erfcxf`?

Reply:
Have updated!

But I believe we will need to use previous test cases, due to tolerance issues. Will that be good? Also, any reason you preferred the `single precision version`?

Thanks!

Reply:
> But I believe we will need to use previous test cases, due to tolerance issues. Will that be good?

Seems that `check_unary_linspace0_1`'s atol 'default' value is too strict. Would you try changing to 1e-6? (It should be also safe for other tests that use the value)

https://github.com/cupy/cupy/blob/462d14e6ebcd4c8b2f89e12b65a92c217ebbb12f/tests/cupyx_tests/scipy_tests/special_tests/test_statistics.py#L31

Reply:
Fixed

Reply:
> EDIT: Ooo, I see. Because of performance reasons!

Yes. Single-precision operations are generally faster on GPU. Can I ask you to implement both single-precision version and double-precision version, instead of providing single-precision version only? We want to use double-precision version when precision matters.

I explain some of `create_ufunc`'s parameters.

```py
log_ndtr = _core.create_ufunc(
    'cupyx_scipy_special_log_ndtr',
    (('f->f', 'out0 = log_ndtr(in0)'), 'd->d'),
    'out0 = log_ndtr(in0)',
    ...
```

The second specifies a list of input and output dtypes supported by the create ufunc. 'f->f' means that it takes a float32 (single-precision) input, then returns a float32 output. 'd->d' means that it takes a float64 (double-precision) input then returns a float64 output.

The third supplies the "default" operation performed by this ufunc. Why I say "default" is that please see the first element of the second parameter `('f->f', 'out0 = log_ndtr(in0)')`.  Here provided a tuple of strings instead of 'f->f' only. This means that, when the input's dtype if float32, the second element of the tuple `'out0 = log_ndtr(in0)'` is used instead of the default operation. It happens to be the same as the default one in your code, but you can give a single-precision version, say, `out0 = log_ndtrf(in0)` here.

Reply:
Interesting! Thanks, for the detailed description!